### PR TITLE
Deprecate legacy cookies behind `LEGACY_COOKIES` profile

### LINKED ISSUE #16181: Deprecate legacy cookie behaviour behind profile
### Description

Since implementing [KEYCLOAK-12125](https://issues.redhat.com/browse/KEYCLOAK-12125), Keycloak adds two cookies with the same value but different `SameSite` settings when the `SameSite=None` is set to none. 

The was due to old browsers not support `SameSite=None`, see https://www.chromium.org/updates/same-site/incompatible-clients/ for details and how to detect them.

Since then, browsers supporting this are wide-spread: https://caniuse.com/?search=SameSite%3DNone (92+%). 

Adding these extra cookies adds to the HTTP headers sent in a response and request, and a default setup with NGINX which limits responses with 4k headers needs to be tweaked to make it work. 

Therefore, I suggest to remove the `*_LEGACY` cookies. Additionally, we should unify our cookie behaviour so that:
- All cookies are `Secure` by default.
- All cookies are `SameSite=Strict`, unless explicitly specified otherwise.

### Motivation

* Simplify Keycloak's code structure, and support ever-green browers.
* Simplify setup of Keycloak behind a loadbalancer

### PR DESCRIPTION
**Pull request by jonkoops**
Closes #16181

### CODE REVIEW

**Review by pedroigor — DISMISSED** [review]

**Review by ahus1 — COMMENTED** [review]
There will be likely some test failures in the Quarkus area for the generated help output with the new option. 

I'd also like to see something in the release notes for this and I'm happy to contribute those.  Anywhere else we would need to document this?

**Comment by ahus1** [comment]
@jonkoops - there's now PR https://github.com/jonkoops/keycloak/pull/13 which fixes "Keycloak CI / Quarkus IT (zip)". The other failing tests are IMHO unstable tests. 

Please squash the commits after merging. Also, please add an issue number to the commit (the original commit for this PR doesn't contain an issue number). 

Thanks!

**Comment by jonkoops** [comment]
Thanks for looking into that @ahus1, I've pulled in your changes and squashed the commits.

**Comment by jonkoops** [comment]
> Also, please add an issue number to the commit (the original commit for this PR doesn't contain an issue number).

I don't believe we're doing this any longer. When squash merging the PR the PR number will be added automatically (see the [commits on main](https://github.com/keycloak/keycloak/commits/main)). From there the issue is already linked to the PR, which will automatically close it.

**Comment by ahus1** [comment]
> I don't believe we're doing this any longer. When squash merging the PR the PR number will be added automatically (see the [commits on main](https://github.com/keycloak/keycloak/commits/main)). From there the issue is already linked to the PR, which will automatically close it.

I see some commits doing this, while other don't - see 9176308d7919c28c8ec64588e1012af97f6b76f5 as an example which doesn't. So different maintainers seem to handle it differently. The CONTRIBUTING.md still asks for it, do you think this is outdated? IMHO these are two things to raise with the maintainers if you don't want to have this in the commit message.

**Comment by mhajas** [comment]
As far as I know, it is still necessary to include `Closes #XXXXXX` in the commit message. I see I merged some commits that do not contain it and the reason is that I didn't check the commit message. I expected that if the PR description contains it, it comes from the commit message.

**Comment by jonkoops** [comment]
I think this discussions is getting off-track to what the contents of this PR is, let's continue this on the mailing list.

**Comment by jonkoops** [comment]
@ahus1 @mhajas I'll make sure to add the correct information to the commit through the GitHub UI  when merging it into main. Could y'all take another look at the PR and leave a review?

**Review by ahus1 — DISMISSED** [review]
Thanks for implementing this. I reviewed it and also did a manual smoke test.

**Comment by jonkoops** [comment]
Thanks for the assist @ahus1. Java is not my native land so I appreciate it.

**Review by stianst — DISMISSED** [review]

**Comment by stianst** [comment]
@vmuzikar could you review this one?

**Review by pedroigor — DISMISSED** [review]

**Review by vmuzikar — CHANGES_REQUESTED** [review]
I'm afraid removing legacy cookies breaks HTTP (non-SSL) requests, at least to the admin console. That is because we have to force `Secure` anytime we use `SameSite=None`:
https://github.com/keycloak/keycloak/blob/eab835baadee6f04ba2082da5a01d7740bd08dfc/services/src/main/java/org/keycloak/services/util/CookieHelper.java#L68
This also concerns the default `kc.sh start-dev` profile.

I think we have to consider using `SameSite=Lax` when HTTP is explicitly enabled and document this behaviour.

**Comment by pedroigor** [comment]
@vmuzikar You're right and I missed it. It works for fine in most browsers (E.g.: FF and Chrome) but only if the domain name is `localhost`. Those browser consider connections to localhost as `secure` so that you don't need a HTTPS channel.

However, this change as is should impact people not only running in dev mode (e.g.: `start-dev`) but also those exposing the server through HTTP when using a domain name other than `localhost`. It is a break change as is.

Ideally, we should decide (instead of hard-coded as it is today) the same site policy (and if it is secure) for cookies depending on both realm and server settings.

**Comment by pedroigor** [comment]
@vmuzikar Looking at it more closely looks like we can rely on the `required ssl` setting at the realm level to decide whether or not `sameSite=None+secure`  should be set.

By default, realms are set with `required ssl=external` but we are forcing `sameSite=None` without taking into account whether the `secure` flag is enabled. It looks like a problem that we need to handle better.

We could avoid setting `None` (and forcing `secure`) if the realm does force TLS. This code [1] does look right because `secure` should not be blindly set if `sameSite=None`. Instead, we need to consider whether the server expects communication over TLS based on the `required ssl` settings and then configure `sameSite`.

In dev mode and using default realm settings (`required ssl=external`) we should not include `sameSite=None` because connections mostly happen through a non-secure channel. Even if you are using a different hostname/domain.

In production mode and using default realm settings, we should not include `sameSite=None` for clients using a local address if requests are made through a non-secure channel.

However, when running behind a proxy, we are going to rely on the client address set (or not) by the proxy. If the proxy is opening connections through a non-secure channel, the server might end up not marking cookies as `secure` even though your application is publicly available through a secure channel.

In other words, it needs a bit more analysis to make sure we are not breaking existing deployments and their assumptions about how we set those cookies.


[1] https://github.com/keycloak/keycloak/blob/eab835baadee6f04ba2082da5a01d7740bd08dfc/services/src/main/java/org/keycloak/services/util/CookieHelper.java#L70

**Comment by vmuzikar** [comment]
@pedroigor I think we can make some assumptions to set `SameSite=None; Secure` or `SameSite=Lax` based on server and realm settings, as you mentioned. However, to be on the safe side, I think we should have a way for the users to override it if necessary as it is a potential blocker (either too loose security in `Lax`, or unusable cookies in `None` and hence unable to log in).

**Comment by pedroigor** [comment]
@vmuzikar Not sure if we need to explicitly set samesite  to `lax` but just rely on the default from browsers?

**Comment by vmuzikar** [comment]
> @vmuzikar Not sure if we need to explicitly set samesite to `lax` but just rely on the default from browsers?

I'm not sure. According to the [docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite), some browsers defaults to `None`.

**Comment by jonkoops** [comment]
This is growing a bit beyond the scope that I can pick up at the moment (have some other high prio tasks). If anyone wants to take this PR let me know.

Personally I think the best approach is always to use the most secure default and document this properly. If users run into issues with these defaults they should opt-out explicitly (e.g. `--no-secure-cookies`).

**Comment by ahus1** [comment]
A note to those who might pick up this PR up later:

There used to be a PR for the docs, which is no longer mergeable due to the move of the docs: https://github.com/keycloak/keycloak-documentation/pull/1761

**Comment by jonkoops** [comment]
I am closing this PR, as it has been out there for too long. I'll try to do a second attempt at this soon.

**Comment by MGLL** [comment]
Hello @jonkoops @pedroigor, 

I'm coming from this [discussion](https://github.com/keycloak/keycloak/discussions/17116) and the topic that the `KEYCLOAK_SESSION` and `KEYCLOAK_SESSION_LEGACY` was flagged from an external Penetration Test (company) as it's not HttpOnly. 

From what I read here and from this [code comment ](https://github.com/keycloak/keycloak/blob/7d3ce4bd9b79aeb0cff4e099a8460334460999c7/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java#L799), it seems as of today, it's not possible to set it as HttpOnly.

So my questions are, 
- what is the risk having this cookie not set to HttpOnly ? (is it an acceptable risk ?)
- is there a plan to allow this cookie as HttpOnly ?

Related discussions:
- https://github.com/keycloak/keycloak/discussions/23352
- https://github.com/keycloak/keycloak/discussions/17116
- https://github.com/keycloak/keycloak/discussions/23338

Best regards

**Comment by jonkoops** [comment]
> what is the risk having this cookie not set to HttpOnly?

I think that OWASP can give you the best [answer to this question](https://owasp.org/www-community/HttpOnly). We need those cookies to be accessible to JavaScript so we can verify the user has an active session, or otherwise redirect the user to restart the login flow.

> is there a plan to allow this cookie as HttpOnly ?

Not at this time. But we're going to be re-evaluating the defaults on cookies in the coming year or so, which may or may not cause this to change.

**Comment by MGLL** [comment]
@jonkoops Thanks a lot for the response. 
So as of today, is there any known risk mitigation regarding `KEYCLOAK_SESSION` ? (or it's just a risk we have to assume.)

**Comment by MGLL** [comment]
@jonkoops thanks again, so I guess that's a risk we have to live with for now. 

From what I understand, this cookie can't be set as HttpOnly because some Keycloak core elements relies on it (Admin Console, User Account, Login flow, etc). 
Which would mean, that to make it HttpOnly and make it secure, it would require to refactor quite some logics on how this elements interact with the user's session.
